rpc/wrtc_base_channel.go

```diff
@@ -61,101 +61,96 @@ func newBaseChannel(
 
 	var connID string
 	var connIDMu sync.Mutex
-	var peerDoneOnce bool
-	doPeerDone := func() {
-		// Cancel base channel context so streams on the channel will stop trying
-		// to receive messages when the peer is done.
-		ch.cancel()
-
-		if !peerDoneOnce && onPeerDone != nil {
-			peerDoneOnce = true
-			onPeerDone()
+	var peerDoneOnce sync.Once
+	peerConn.OnICEConnectionStateChange(func(connectionState webrtc.ICEConnectionState) {
+		if ch.closed.Load() {
+			return
 		}
-	}
-	connStateChanged := func(connectionState webrtc.ICEConnectionState) {
-		ch.activeBackgroundWorkers.Add(1)
-
-		utils.PanicCapturingGo(func() {
-			defer ch.activeBackgroundWorkers.Done()
 
-			if ch.closed.Load() {
-				doPeerDone()
+		switch connectionState {
+		case webrtc.ICEConnectionStateDisconnected,
+			webrtc.ICEConnectionStateFailed,
+			webrtc.ICEConnectionStateClosed:
+
+			// The Disconnected/Failed/Closed states are all interpreted as terminal. In theory,
+			// WebRTC allows for moving from the disconnected -> connected state if the network
+			// heals or new candidates are presented. We forgo that optimization of preserving an
+			// existing PeerConnection object and instead opt for clients to redial and perform a
+			// new round of signaling/answering.
+			connIDMu.Lock()
+			currConnID := connID
+			connIDMu.Unlock()
+			if currConnID == "" { // make sure we've gathered information before
```

Comment:
> // make sure we've gathered information before

I know this was preexisting, but is this comment missing something? I don't understand what it means or what it's referring to.

Reply:
I agree the comment is bad. I added in why* connID would be empty/non-empty that gives a better picture to the intent.